File: leetcode/solution_39.py
'''
Summary

Attempt #1
Your Own answer?: Yes
Time spent: 30m
Time Complexity: ?
Runtime: 145 ms, faster than 16.30% of Python3 online submissions for Combination Sum.
Space Complexity: O(N) where N is target
Memory Usage: 14.3 MB, less than 52.37% of Python3 online submissions for Combination Sum.
'''


import sys
from typing import List
input = sys.stdin.readline

class Solution:
    def __init__(self) -> None:
        candidates = [2,3,6,7]
        target = 7
        print(self.combinationSum(candidates, target))

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        size = len(candidates)
        ans = []

        def dfs(start, path, goal):
            if goal < 0:
                return
            elif goal == 0:
                ans.append(path)
                return
            for i in range(start, size):
                dfs(i, path + [candidates[i]], goal - candidates[i])

        dfs(0, [], target)
        return ans

Solution()

File: leetcode/solution_39.cs
// Time Complexity: ??
// Space Complexity: ??

public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        var result = new List<IList<int>>();

        void dfs(int index, int newTarget, List<int> accumulated)
        {
            if (newTarget == 0)
            {
                result.Add(accumulated);
                return;
            }
            
            for (int i = index; i < candidates.Length; i++)
            {
                var candidate = candidates[i];
                if (candidate <= newTarget)
                {
                    var newList = new List<int>(accumulated);
                    newList.Add(candidate);
                    dfs(i, newTarget - candidate, newList);
                }
            }
        }

        dfs(0, target, new List<int>());

        return result;
    }
}